qiskit_addon_utils/exp_vals/expectation_values.py

```diff
@@ -0,0 +1,255 @@
+# This code is a Qiskit project.
+#
+# (C) Copyright IBM 2025.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Utility functions for calculating expectation values of observables."""
+
+from __future__ import annotations
+
+import numpy as np
+from qiskit.primitives import BitArray
+from qiskit.quantum_info import Pauli, SparseObservable, SparsePauliOp
+from qiskit.primitives.containers.observables_array import ObservablesArray
+
+def expectation_values(
+    bool_array: np.ndarray[np._bool],
+    basis_dict: dict[Pauli, list[SparsePauliOp | None]],
+    meas_basis_axis: int,
+    avg_axis: int | tuple[int] | None = None,
+    meas_flips: np.ndarray[np._bool] | None = None,
+    pec_signs: np.ndarray[np._bool] | None = None,
+    postselect_mask: np.ndarray[np._bool] | None = None,
+):
+    """Computes expectation values from boolean data.
+
+    Uses data in `bool_array`, acquired with measurement bases as ordered in keys of `basis_dict`, to compute observables encoded in values of `basis_dict`.
```

Comment:
This can be dangerous... While Python dict objects have been order-preserving for a while, that still can depend on their construction (e.g. dict comprehension where the keys are taken from a set (which is unordered) results in unpredictably ordered dict keys). While that is a pathological edge case, if there is any way that we can avoid this, I'd prefer for a safer route.

Reply:
Makes sense. One way or another though, the user will have to provide the ordering to match the ordering of data in `bool_array`. We could reduce the ambiguity by just taking this input as a `list` instead of a `dict` (since we don't currently use the keys anyway).